byteps/common/thread_pool.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copy From https://github.com/progschj/ThreadPool/blob/master/ThreadPool.h
+ */
+#ifndef THREAD_POOL_H
+#define THREAD_POOL_H
+
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <stdexcept>
+#include <thread>
+#include <vector>
+
+class ThreadPool {
+ public:
+  ThreadPool(size_t);
+  template <class F, class... Args>
+  void enqueue(F&& f, Args&&... args);
+  ~ThreadPool();
+
+ private:
+  // need to keep track of threads so we can join them
+  std::vector<std::thread> workers;
+  // the task queue
+  std::queue<std::function<void()> > tasks;
+
+  // synchronization
+  std::mutex queue_mutex;
+  std::condition_variable condition;
+  bool stop;
+};
+
+// the constructor just launches some amount of workers
+inline ThreadPool::ThreadPool(size_t threads) : stop(false) {
+  for (size_t i = 0; i < threads; ++i)
+    workers.emplace_back([this] {
+      for (;;) {
+        std::function<void()> task;
+
+        {
+          std::unique_lock<std::mutex> lock(this->queue_mutex);
+          this->condition.wait(
+              lock, [this] { return this->stop || !this->tasks.empty(); });
+          if (this->stop && this->tasks.empty()) return;
+          task = std::move(this->tasks.front());
+          this->tasks.pop();
+        }
+
+        task();
+      }
+    });
+}
+
+// add new work item to the pool
+template <class F, class... Args>
+void ThreadPool::enqueue(F&& f, Args&&... args) {
+  {
+    std::unique_lock<std::mutex> lock(queue_mutex);
+
+    // don't allow enqueueing after stopping the pool
+    if (stop) throw std::runtime_error("enqueue on stopped ThreadPool");
+
+    tasks.emplace([f]() { f(); });
+  }
+  condition.notify_one();
+}
+
```

Comment:
1. You are not using `Args&& ...` parameter pack.
2. "tasks.emplace([f]() { f(); });": As you are playing with universal reference, do not use capture-by-copy. Or it will involve a copy.
![image](https://user-images.githubusercontent.com/38074777/87179435-4f44f100-c311-11ea-9250-d26748fcc0ef.png)
Write sth. like `[invoke = std::forward<F>(f)]{ invoke(); }();`

3.Your `decltype(tasks)` is `Container<std::function>`, therefore I think there's no need to re-wrap a lambda.

```c++
template <class F>
void ThreadPool::enqueue(F&& f) {
  {
    std::lock_guard<std::mutex> lock(queue_mutex);

    if (stop) throw std::runtime_error("enqueue on stopped ThreadPool");

    tasks.emplace(std::forward<F>(f));
  }
  condition.notify_one();
}
```

Reply:
Thanks for your suggestions. I'll have a look.

Reply:
I've made my last comment a suggested change.

Reply:
LGTM. I will merge it.

Reply:
it incurs a compilation error.
```sh
byteps/common/thread_pool.h:58:6: error: prototype for ‘void ThreadPool::enqueue(F&&)’ does not match any in class ‘ThreadPool’
 void ThreadPool::enqueue(F&& f) {
      ^~~~~~~~~~
byteps/common/thread_pool.h:21:8: error: candidate is: template<class F, class ... Args> void ThreadPool::enqueue(F&&, Args&& ...)
   void enqueue(F&& f, Args&&... args);
```

Reply:
This is because you have a function definition before this function implementation. I will modify this in a wider range of codes.